File: android/src/main/kotlin/com/neutronstarer/npcx/NpcxPlugin.kt
package com.neutronstarer.npcx

import android.os.Handler
import android.os.Looper
import androidx.annotation.NonNull
import com.neutronstarer.npc.*
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.JSONMessageCodec
import org.json.JSONObject
import java.nio.ByteBuffer

/** NpcxPlugin */
class NpcxPlugin: NPC(null), FlutterPlugin {

  companion object {
    fun of(engine: FlutterEngine?): NpcxPlugin? {
      return engine?.plugins?.get(NpcxPlugin::class.java) as? NpcxPlugin
    }
  }

  private lateinit var name: String
  private lateinit var messenger : BinaryMessenger
  private val codec = JSONMessageCodec.INSTANCE
  private fun didReceive(bytes: ByteBuffer?) {
    try {
      val message = codec.decodeMessage(bytes) as? JSONObject ?: return
      val rawValue = message["typ"] as? Int ?: return
      val typ = Typ.fromRawValue(rawValue)
      val id = message.opt("id") as? Int
      val method = message.opt("method") as? String
      val param = message.opt("param")
      val error = message.opt("error")
      receive(Message(typ = typ, id = id, method = method, param = param, error = error))
    }finally {

    }

  }

  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    name = "com.neutronstarer.npcx"
    messenger = flutterPluginBinding.binaryMessenger
    messenger.setMessageHandler(name) { message, _ ->
      didReceive(bytes = message)
    }
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    cleanUpDeliveries("disconnected")
  }

  override fun send(message: Message) {
    super.send(message)
    val map = mutableMapOf<String, Any>()
    val typ = message.typ.rawValue
    map["typ"] = typ
    val id = message.id
    if (id != null){
      map["id"] = id
    }
    val method = message.method
    if (method != null){
      map["method"] = method
    }
    val param = message.param
    if (param != null){
      map["param"] = param
    }
    val error = message.error
    if (error != null){
      map["error"] = error
    }
    val bytes = codec.encodeMessage(map)
    Handler(Looper.getMainLooper()).post {
      messenger.send(name, bytes)
    }
  }

}

File: ios/Classes/NpcxPlugin.swift
import Flutter
import Foundation
import NPC

fileprivate struct AssociatedKeys {
    static var npcx: Void?
}

public class NpcxPlugin: NPC, FlutterPlugin {
    
    public static func of(_ registry: FlutterPluginRegistry) -> NpcxPlugin? {
        if let registry = registry as? FlutterAppDelegate {
            return registry.npcx()
        }
        if let registry = registry as? FlutterViewController {
            return registry.npcx()
        }
        if let registry = registry as? FlutterEngine {
            return registry.npcx()
        }
        return nil
    }
    
    private let name: String
    private weak var messenger: FlutterBinaryMessenger?
    private let codec = FlutterJSONMessageCodec.sharedInstance()
    
    private init(_ name: String, _ messenger: FlutterBinaryMessenger){
        self.name = name
        self.messenger = messenger
        super.init(nil)
        messenger.setMessageHandlerOnChannel(name) {[weak self] data, reply in
            guard let self = self, let message = self.codec.decode(data) as? [String: Any], let rawValue = message["typ"] as? Int, let typ = Typ(rawValue: rawValue) else {
                return
            }
            self.receive(Message(typ: typ, id: (message["id"] as? Int) ?? 0, method: message["method"] as? String, param: message["param"], error: message["error"]))
        }
    }
    
    public override func send(_ message: Message) {
        var map = [String: Any]()
        map["typ"] = message.typ.rawValue
        map["id"] = message.id
        if let method = message.method {
            map["method"] = method
        }
        if let param = message.param {
            map["param"] = param
        }
        if let error = message.error {
            map["error"] = error
        }
        let msg = codec.encode(map)
        messenger?.send(onChannel: name, message: msg)
    }
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        // Do not use channel for preventing circular references
        let name = "com.neutronstarer.npcx"
        let messenger = registrar.messenger()
        let instance = NpcxPlugin(name, messenger)
        objc_setAssociatedObject(messenger, &AssociatedKeys.npcx, instance, .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
    }
}


extension FlutterAppDelegate {
    func npcx() -> NpcxPlugin? {
        return (self.perform(NSSelectorFromString("rootFlutterViewController")).takeUnretainedValue() as? FlutterViewController)?.npcx()
    }
}

@objc public extension FlutterViewController {
    func npcx() -> NpcxPlugin? {
        return engine?.npcx()
    }
}

@objc public extension FlutterEngine {
    func npcx() -> NpcxPlugin? {
        let messenger = binaryMessenger
        return objc_getAssociatedObject(messenger, &AssociatedKeys.npcx) as? NpcxPlugin
    }
}

